Memory QOS Alpha2 KEP 2570 blog post (#39853)
kubernetes/website: content/en/blog/_posts/2023-05-05-memory-qos-cgroups-v2/index.md

  specify the hard limit for memory usage. If the memory consumption goes above this
  level, the kernel invokes its OOM Killer.

  cgroups v2 also added `memory.high` configuration . Memory QoS uses `memory.high`
  to set memory usage throttle limit. If the `memory.high` limit is breached,
  the offending cgroups are throttled, and the kernel tries to reclaim memory
  which may avoid an OOM kill.

## How it works

### Cgroups v2 memory controller interfaces & Kubernetes container resources mapping

Memory QoS uses the memory controller of cgroups v2 to guarantee memory resources in
Kubernetes. cgroupv2 interfaces that this feature uses are: 
* `memory.max`
* `memory.min`
* `memory.high`.

{{< figure src="/blog/2023/05/05/qos-memory-resources/memory-qos-cal.svg" title="Memory QoS Levels" alt="Memory QoS Levels" >}}

`memory.max` is mapped to memory limit specified in the Pod spec. The kubelet and 
the container runtime enforce the limit and prevent the container from using more
than the configured resource limit. If a process in a container tries to consume
more than the specified limit, kernel terminates the process with an out of
memory Out of Memory (OOM) error. 

> {{< figure src="/blog/2023/05/05/qos-memory-resources/container-memory-max.svg" title="`memory.max` maps to `limits.memory`" alt="`memory.max` maps to `limits.memory`" >}}

Reviewer: I'm afraid you can't use Markdown in captions.

`memory.min` is mapped to memory `requests`, which results in reservation of memory resources
that can never reclaimed by the kernel. This is how Memory QoS ensures the availability of 
memory for Kubernetes pods. If there's no unprotected reclaimable memory available, OOM 
killer is invoked.

> {{< figure src="/blog/2023/05/05/qos-memory-resources/container-memory-min.svg" title="`memory.min` maps to `requests.memory`" alt="`memory.min` maps to `requests.memory`" >}}

Reviewer: I'm afraid you can't use Markdown in captions.

For memory protection, in addition to the original way of limiting memory usage, Memory QoS
throttles workload approaching its memory limit, ensuring that the system is not overwhelmed
by sporadic increases in memory usage. A new field, `memoryThrottlingFactor`, is available in
the KubeletConfiguration when you enable MemoryQoS feature. It is set to 0.9 by default. 
`memory.high` is mapped to throttling limit calculated by using `memoryThrottlingFactor`,
`spec.containers[].resources.requests.memory` and `spec.containers[].resources.limits.memory`
as in the formula below, and rounding down the value to the nearest page size:

> {{< figure src="/blog/2023/05/05/qos-memory-resources/container-memory-high.svg" title="`memory.high` formula" alt="`memory.high` formula" >}} 

Reviewer: I'm afraid you can't use Markdown in captions.

> **Note**: If a container has no memory limits specified, we substitute `limits.memory` for node allocatable memory.

Reviewer: Can you rephrase to [avoid using “we”](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-using-we)?

        <br>
        <i>It is mapped to the container's memory request specified in Pod manifest..</i>
        </td>
   </tr>
   <tr>
       <td>memory.high</td>
       <td><code>memory.high</code> specifies the memory usage throttle limit. 
       This is the main mechanism to control a cgroup's memory use. If 
       cgroups memory use goes over the high boundary specified here,
       the cgroups processes are throttled and put under heavy reclaim pressure.
       <br>
       <br>
       <i>Kubernetes uses a formula to calculate <code>memory.high</code>,
       depending on container's memory request, memory limit or node allocatable memory
      (if container's memory limit is empty) and a throttling factor.
      Please refer to the <a href="https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2570-memory-qos">KEP</a>
      for more details on the formula.</i>
       </td>
   </tr>
</table>

**Note** `memory.high` is set only on container level cgroups while `memory.min` is set on
container, pod and node level cgroups.

### `memory.min` calculations for cgroups heirarchy

When container memory requests are made, kubelet passes `memory.min` to the back-end 

> CRI runtime (possibly containerd, cri-o) via the `Unified` field in CRI during 

Reviewer: Even better:
nit
```suggestion
CRI runtime (such as containerd or CRI-O) via the `Unified` field in CRI during 
```

container creation. The `memory.min` in container level cgroups will be set to:  

$memory.min =  pod.spec.containers[i].resources.requests[memory]$  
<sub>for every i<sup>th</sup> container in a pod</sub>

Since the `memory.min` interface requires that the ancestor cgroups directories are all 
set, the pod and node cgroups directories need to be set correctly. 

`memory.min` in pod level cgroup:  
$memory.min = \sum_{i=0}^{no. of pods}pod.spec.containers[i].resources.requests[memory]$  

`memory.min` in node level cgroup:  

> $memory.min = \sum_{i}^{no. of nodes}\sum_{j}^{no. of pods}pod[i].spec.containers[j].resources.requests[memory]$  

Reviewer: This might be easier to read as an SVG formula with alt text.

Author: looks fine to me in the preview. Also, it will make it look more consistent to either use SVG for all, or use markdown foll all. Attaching ss. Thoughts? 
![Screenshot 2023-04-17 at 11 49 29 AM](https://user-images.githubusercontent.com/5120675/232582889-b1898271-d0d8-49e7-afe9-144bd2945eb3.png)

Kubelet will manage the cgroups hierarchy of the pod level and node level cgroups 
directly using the libcontainer library (from the runc project), while container
cgroups limits are managed by the container runtime.

### Support for Pod QoS classes

Based on user feedback for the Alpha feature in Kubernetes v1.22, some users would like 
to opt out of MemoryQoS on a per-pod basis to ensure there is no early memory throttling.
Therefore, in Kubernetes v1.27 Memory QOS also supports memory.high to be set as per 
Quality of Service(QoS) for Pod classes. Following are the different cases for memory.high
as per QOS classes:

1. **Guaranteed pods** by their QoS definition require memory requests=memory limits and are 
not overcommitted. Hence MemoryQoS feature is disabled on those pods by not setting 
memory.high. This ensures that Guaranteed pods can fully use their memory requests up
to their set limit, and not hit any throttling.

2. **Burstable pods** by their QoS definition require at least one container in the Pod with
CPU or memory request or limit set.
    
   * When requests.memory and limits.memory are set, the formula is used as-is:

>      {{< figure src="/blog/2023/05/05/qos-memory-resources/container-memory-high-limit.svg" title="`memory.high` when requests and limits are set" alt="`memory.high` when requests and limits are set" >}}

Reviewer: I'm afraid you can't use Markdown in captions.

>     * When requests.memory and limits.memory are set, the formula is used as-is:
> 
>     {{< figure src="/blog/2023/05/05/qos-memory-resources/container-memory-high-limit.svg" title="`memory.high` when requests and limits are set" alt="`memory.high` when requests and limits are set" >}}
> 
>     * When requests.memory is set, limits.memory is not set, we substitute limits.memory
>     for node allocatable memory in the formula:
> 
>     {{< figure src="/blog/2023/05/05/qos-memory-resources/container-memory-high-no-limits.svg" title="`memory.high` when requests and limits are not set" alt="`memory.high` when requests and limits are not set" >}}

Reviewer 2: ```suggestion
   * When requests.memory and limits.memory are set, the formula is used as-is:

     {{< figure src="/blog/2023/05/05/qos-memory-resources/container-memory-high-limit.svg" title="`memory.high` when requests and limits are set" alt="`memory.high` when requests and limits are set" >}}

   * When requests.memory is set, limits.memory is not set, we substitute limits.memory
     for node allocatable memory in the formula:

     {{< figure src="/blog/2023/05/05/qos-memory-resources/container-memory-high-no-limits.svg" title="`memory.high` when requests and limits are not set" alt="`memory.high` when requests and limits are not set" >}}
```

Reviewer 2: be careful about the indentation. 4-spaces indentation sometimes confuses the parser. The parser may treat it as code snippets. For numbered lists, please always indent by 3-spaces.

>      {{< figure src="/blog/2023/05/05/qos-memory-resources/container-memory-high-no-limits.svg" title="`memory.high` when requests and limits are not set" alt="`memory.high` when requests and limits are not set" >}}

Reviewer: I'm afraid you can't use Markdown in captions.

> 3. **BestEffort** by their QoS definition do not require any memory or CPU limits or requests.
> For this case, kubernetes sets requests.memory = 0 and substitute limits.memory for node allocatable
> memory in the formula:
> 
>     {{< figure src="/blog/2023/05/05/qos-memory-resources/container-memory-high-best-effort.svg" title="`memory.high` for BestEffort Pod" alt="memory.high` for BestEffort Pod" >}}

Reviewer 2: ```suggestion
3. **BestEffort** by their QoS definition do not require any memory or CPU limits or requests.
   For this case, kubernetes sets requests.memory = 0 and substitute limits.memory for node allocatable
   memory in the formula:

   {{< figure src="/blog/2023/05/05/qos-memory-resources/container-memory-high-best-effort.svg" title="`memory.high` for BestEffort Pod" alt="memory.high` for BestEffort Pod" >}}
```

>    {{< figure src="/blog/2023/05/05/qos-memory-resources/container-memory-high-best-effort.svg" title="`memory.high` for BestEffort Pod" alt="`memory.high` for BestEffort Pod" >}}

Reviewer: I'm afraid you can't use Markdown in captions.

> **Summary**: Only Pods in Burstable and Best effort QoS classes will set memory.high.

Reviewer 2: ```suggestion
**Summary**: Only Pods in Burstable and BestEffort QoS classes will set `memory.high`.
```

> Guaranteed QoS pods do not set memory.high as their memory is guaranteed.

Reviewer 2: ```suggestion
Guaranteed QoS pods do not set `memory.high` as their memory is guaranteed.
```

## How do I use it?

The prerequisites for enabling Memory QoS feature on your Linux node are:

1. Verify the [requirements](/docs/concepts/architecture/cgroups/#requirements) 

> related to [Kubernetes support for cgroups v2](/docs/concepts/architecture/cgroups)
> are met. 

Reviewer 2: ```suggestion
   related to [Kubernetes support for cgroups v2](/docs/concepts/architecture/cgroups)
   are met. 
```

> 2. Ensure CRI Runtime supports Memory QoS. In Alpha phase of this feature,
>    only containerd and cri-o are supported. Related PRs:

Reviewer: ```suggestion
1. Ensure CRI Runtime supports Memory QoS. At the time of writing, only
   containerd and CRI-O provide support compatible with this alpha. That was
   implemented in the following PRs:
```

Kubernetes doesn't provide support for runtimes' features; mostly I see it as CRI implementations providing support for Kubernetes features. The CRI is defined by _this_ project, and runtime adapters can choose to conform to it.

> only containerd and cri-o are supported. Related PRs:
>     * Containerd: [Feature: containerd-cri support LinuxContainerResources.Unified #5627](https://github.com/containerd/containerd/pull/5627).
>     * CRI-O: [implement kube alpha features for 1.22 #5207](https://github.com/cri-o/cri-o/pull/5207).

Reviewer 2: ```suggestion
   only containerd and cri-o are supported. Related PRs:
   * Containerd: [Feature: containerd-cri support LinuxContainerResources.Unified #5627](https://github.com/containerd/containerd/pull/5627).
   * CRI-O: [implement kube alpha features for 1.22 #5207](https://github.com/cri-o/cri-o/pull/5207).
```
